src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.DecisionDagRewriter.cs

```diff
@@ -865,59 +870,237 @@ private void EnsureStringHashFunction(int labelsCount, SyntaxNode syntaxNode)
             {
-                // This node is used even when there is no when clause, to record bindings. In the case that there
-                // is no when clause, whenClause.WhenExpression and whenClause.WhenFalse are null, and the syntax for this
-                // node is the case clause.
-
-                // We need to assign the pattern variables in the code where they are in scope, so we produce a branch
-                // to the section where they are in scope and evaluate the when clause there.
-                var whenTrue = (BoundLeafDecisionDagNode)whenClause.WhenTrue;
-                LabelSymbol labelToSectionScope = GetDagNodeLabel(whenClause);
-
-                ArrayBuilder<BoundStatement> sectionBuilder = BuilderForSection(whenClause.Syntax);
-                sectionBuilder.Add(_factory.Label(labelToSectionScope));
-                foreach (BoundPatternBinding binding in whenClause.Bindings)
-                {
-                    BoundExpression left = _localRewriter.VisitExpression(binding.VariableAccess);
-                    // Since a switch does not add variables to the enclosing scope, the pattern variables
-                    // are locals even in a script and rewriting them should have no effect.
-                    Debug.Assert(left.Kind == BoundKind.Local && left == binding.VariableAccess);
-                    BoundExpression right = _tempAllocator.GetTemp(binding.TempContainingValue);
-                    if (left != right)
+                // The way the DAG is prepared, it is possible for different `BoundWhenDecisionDagNode` nodes to
+                // share the same `WhenExpression` (same `BoundExpression` instance).
+                // So we can't just lower each `BoundWhenDecisionDagNode` separately, as that would result in duplicate blocks
+                // for the same `WhenExpression` and such expressions might contains labels which must be emitted once only.
+
+                // For a simple `BoundWhenDecisionDagNode` (with a unique `WhenExpression`), we lower to something like:
+                //   labelToSectionScope;
+                //   if (... logic from WhenExpression ...)
+                //   {
+                //     jump to whenTrue label
+                //   }
+                //   jump to whenFalse label
+
+                // For a complex `BoundWhenDecisionDagNode` (where the `WhenExpression` is shared), we lower to something like:
+                //   labelToSectionScope;
+                //   whenNodeIdentifierLocal = whenNodeIdentifier;
+                //   goto labelToWhenExpression;
+                //
+                // and we'll also create a section for the shared `WhenExpression` logic:
+                //   labelToWhenExpression;
+                //   if (... logic from WhenExpression ...)
+                //   {
+                //     switch on whenNodeIdentifierLocal with dispatches to whenTrue labels
```

Comment:
I don't think the success path needs a dispatch, duplicated when-clauses will always jump to the same label if true.

Reply:
Indeed. Thanks